version: '3.8'

services:
  database:
    container_name: bookrestapi_database
    image: postgres:15
    restart: on-failure
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
    ports:
      - "5432:5432"
    volumes:
      - ./db/data:/var/lib/postgresql/data
      - ./db/scripts/:/docker-entrypoint-initdb.d/
    networks:
      - bookrestapi_net
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 10s
      timeout: 5s
      retries: 5

  server:
    container_name: bookrestapi_server  
    build: .
    depends_on:
      database:
        condition: service_healthy
    restart: on-failure
    networks:
      - bookrestapi_net
    ports:
      - "8080:8080"
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8080/health"]
      interval: 10s
      timeout: 5s
      retries: 3

volumes:
  bookrestapi_data:
    driver: local

networks:
  bookrestapi_net:
    driver: bridge
